src/bootloaders/systemd-class.c

```diff
@@ -51,60 +52,149 @@ static BootLoaderConfig *sd_config = NULL;
+}
+
 bool sd_class_init(const BootManager *manager, BootLoaderConfig *config)
 {
         char *base_path = NULL;
         char *efi_dir = NULL;
         char *vendor_dir = NULL;
         char *entries_dir = NULL;
         char *efi_blob_source = NULL;
-        char *efi_blob_dest = NULL;
-        char *default_path_efi_blob = NULL;
         char *loader_config = NULL;
         const char *prefix = NULL;
+        NcHashmap *copy_pairs = NULL;
+        autofree(char) *shim_path = NULL;
+        bool did_put = false;
 
         sd_config = config;
 
+        /* Init copy pairs table */
+        copy_pairs = nc_hashmap_new_full(nc_string_hash, nc_string_compare, free, free);
+        if (!copy_pairs) {
+                DECLARE_OOM();
+                return false;
+        }
+        sd_class_config.copy_pairs = copy_pairs;
+
         /* Cache all of these to save useless allocs of the same paths later */
         base_path = boot_manager_get_boot_dir((BootManager *)manager);
         OOM_CHECK_RET(base_path, false);
         sd_class_config.base_path = base_path;
 
+        /* EFI Boot directory base */
         efi_dir = nc_build_case_correct_path(base_path, "EFI", "Boot", NULL);
         OOM_CHECK_RET(efi_dir, false);
         sd_class_config.efi_dir = efi_dir;
 
+        /* Our vendor directory base */
         vendor_dir = nc_build_case_correct_path(base_path, "EFI", sd_config->vendor_dir, NULL);
         OOM_CHECK_RET(vendor_dir, false);
         sd_class_config.vendor_dir = vendor_dir;
 
+        /* loader/entries */
         entries_dir = nc_build_case_correct_path(base_path, "loader", "entries", NULL);
         OOM_CHECK_RET(entries_dir, false);
         sd_class_config.entries_dir = entries_dir;
 
         prefix = boot_manager_get_prefix((BootManager *)manager);
 
-        /* EFI paths */
+        /* Determine if we have secure-boot support */
+        shim_path = string_printf("%s/usr/share/shim/shim.efi", prefix);
+        sd_class_config.secure_boot = nc_file_exists(shim_path);
+
+        /* Our main "blob" */
         efi_blob_source =
             string_printf("%s/%s/%s", prefix, sd_config->efi_dir, sd_config->efi_blob);
         sd_class_config.efi_blob_source = efi_blob_source;
 
-        efi_blob_dest = nc_build_case_correct_path(sd_class_config.base_path,
+        /* Main vendor blob - always installed */
+        did_put = sd_class_config_put_copy(NULL,
+                                           efi_blob_source,
+                                           "EFI",
+                                           sd_config->vendor_dir,
+                                           sd_config->efi_blob,
+                                           NULL);
+
+        /* Install secure-boot as primary, ourselves as stage2 */
+        if (sd_class_config.secure_boot) {
+                /* TODO: Add *another* blob to be used by EFI variables by path, not just default */
+                did_put = sd_class_config_put_copy(prefix,
```

Comment:
Ah good point -_- OK yeah leave off the secure boot commit for now - given what we discussed earlier.